### PR TITLE
APL-1842 Implement scan recovery

### PR DESCRIPTION
**Pull request by AndrewBoyarsky**
After applying this pr, a scan will store its own state in db and will recover from this state at startup when it was interrupted.
**Note:** added workaround for catching weld shutdown event and **NOT** fail scan in that case

### CODE REVIEW

**Review by alzinchenko — APPROVED** [review]

**Review by andrii-firstbridge — APPROVED** [review]

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 1]
```diff
@@ -901,7 +901,13 @@ protected int update(int nextUpdate) {
             case 352:
                 apply("ALTER TABLE transaction ALTER COLUMN signature VARBINARY NULL DEFAULT NULL");
             case 353:
-                return 353;
+                apply("ALTER TABLE scan ADD current_height INT NOT NULL DEFAULT 0");
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java:904]: Don't forget to move that SQL into new FlyWay scripts on develop.

Comment (AndrewBoyarsky) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java:904]: Already done, pr is on the way

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/service/blockchain/BlockchainProcessorImpl.java [hunk 2]
```diff
@@ -1457,38 +1420,43 @@ private void scan(int height, boolean validate, boolean shutdown) {
                                 dataSource.rollback(false);
                                 log.debug(e.toString(), e);
                                 log.debug("Applying block " + Long.toUnsignedString(currentBlockId) + " at height "
-                                    + (currentBlock == null ? 0 : currentBlock.getHeight()) + " failed, deleting from database");
-                                Block lastBlock = blockchain.deleteBlocksFrom(currentBlockId);
-                                blockchain.setLastBlock(lastBlock);
-                                popOffToCommonBlock(lastBlock);
+                                    + currentBlock.getHeight() + " failed during scan");
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/service/blockchain/BlockchainProcessorImpl.java:1423]: May be better rewrite to log.debug("{} {}", .....)

**Review by yuriy-larin-firstbridge — APPROVED** [review]

**Review by oleksiylukin — APPROVED** [review]

**Comment by AndrewBoyarsky** [comment]
Merged and adapted to the develop branch. Won't merge

**Pull request closed without merging**